pex/scie/science.py

```diff
@@ -0,0 +1,315 @@
+# Copyright 2024 Pex project contributors.
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import os.path
+import re
+import shutil
+import subprocess
+from collections import OrderedDict
+from subprocess import CalledProcessError
+
+from pex.atomic_directory import atomic_directory
+from pex.common import is_exe, pluralize, safe_mkdtemp, safe_open
+from pex.compatibility import shlex_quote
+from pex.exceptions import production_assert
+from pex.fetcher import URLFetcher
+from pex.hashing import Sha256
+from pex.pep_440 import Version
+from pex.pex_info import PexInfo
+from pex.scie.model import ScieConfiguration, ScieInfo, SciePlatform, ScieStyle, ScieTarget
+from pex.third_party.packaging.version import InvalidVersion
+from pex.typing import TYPE_CHECKING, cast
+from pex.util import CacheHelper
+from pex.variables import ENV, Variables, unzip_dir_relpath
+
+if TYPE_CHECKING:
+    from typing import Any, Dict, Iterator, Optional, cast
+
+    import attr  # vendor:skip
+    import toml  # vendor:skip
+else:
+    from pex.third_party import attr, toml
+
+
+@attr.s(frozen=True)
+class Manifest(object):
+    target = attr.ib()  # type: ScieTarget
+    path = attr.ib()  # type: str
+
+    def binary_name(self, binary_name):
+        # type: (str) -> str
+        return self.target.platform.binary_name(binary_name)
+
+    def qualified_binary_name(self, binary_name):
+        # type: (str) -> str
+        return self.target.platform.qualified_binary_name(binary_name)
+
+
+MIN_SCIENCE_VERSION = Version("0.3.0")
+PTEX_VERSION = "1.1.1"
+SCIE_JUMP_VERSION = "1.1.1"
+
+
+def create_manifests(
+    configuration,  # type: ScieConfiguration
+    name,  # type: str
+    pex_info,  # type: PexInfo
+):
+    # type: (...) -> Iterator[Manifest]
+
+    pex_root = "{scie.bindings}/pex_root"
+    if pex_info.venv:
+        # We let the configure-binding calculate the venv dir at runtime since it depends on the
+        # interpreter executing the venv PEX.
+        installed_pex_dir = ""
+    else:
+        production_assert(pex_info.pex_hash is not None)
+        pex_hash = cast(str, pex_info.pex_hash)
+        installed_pex_dir = os.path.join(pex_root, unzip_dir_relpath(pex_hash))
+
+    env_replace = {
+        "PEX_ROOT": pex_root,
+    }
+    env = {
+        "remove_re": {"PEX_.*"},
+        "replace": env_replace,
+    }
+
+    lift = {
+        "name": name,
+        "ptex": {
+            "id": "ptex",
+            "version": PTEX_VERSION,
+            "argv1": "{scie.env.PEX_BOOTSTRAP_URLS={scie.lift}}",
+        },
+        "scie_jump": {"version": SCIE_JUMP_VERSION},
+        "files": [{"name": "configure-binding.py"}, {"name": "pex"}],
+        "commands": [
+            {
+                "env": env,
+                "exe": "{scie.bindings.configure:PYTHON}",
+                "args": ["{scie.bindings.configure:PEX}"],
+            }
+        ],
+        "bindings": [
```

Comment:
This is important for perf. Its pretty bad to package up your "native Python executable" and see it take ~70ms (for cowsay) when a plain `--venv --sh-boot` gets ~20ms. The binding gets any PEX that is scie'd up to `--sh-boot` perf levels.